app/javascript/src/ChangePassword/components/ChangePassword.jsx

```diff
@@ -18,25 +18,111 @@ type Props = {
-    length: { minimum: 1 }
+    length: { minimum: 1 },
+    equality: PASSWORD
   }
 }
 
-const ChangePassword = ({ lostPasswordToken, url, errors }: Props) => {
-  const [isFormDisabled, setIsFormDisabled] = useState(true)
+const isFormValid = (formNode) => validateForm(formNode, validationConstraints)
+
+const useFormState = (formNode) => {
+  const [isDisabled, setIsDisabled] = useState(true)
+
+  const onChange = () => {
+    const errors = isFormValid(formNode)
+    setIsDisabled(!!errors)
+  }
+
+  return {
+    isDisabled,
+    onChange
+  }
+}
+
+const usePasswordState = () => {
+  const [value, setValue] = useState('')
+  const [isValid, setIsValid] = useState(undefined)
+
+  const onChange = (value, event) => {
+    setValue(value)
+    setIsValid(!!validateSingleField(event))
+  }
 
-  const onChange = (event) => {
-    const errors = validateForm(event.currentTarget, validationConstraints)
-    setIsFormDisabled(!!errors)
+  const onBlur = (event) => {
+    const errors = validateSingleField(event)
+    setIsValid(!!errors)
   }
 
+  return {
+    value,
+    isValid,
+    onChange,
+    onBlur
+  }
+}
+
+const getPasswordConfirmationErrors = (formNode) => {
+  const errors = isFormValid(formNode)
+  const passwordConfirmationError = (errors && errors[PASSWORD_CONFIRMATION]) && errors[PASSWORD_CONFIRMATION][0]
+  const passwordsMatchError = passwordConfirmationError && passwordConfirmationError.includes('is not equal to')
+  return {
+    hasError: !!passwordConfirmationError,
+    errorMessage: passwordsMatchError ? 'mustMatch' : 'isMandatory'
+  }
+}
+
+const compareInputsLength = (value) => {
+  const passwordInput = document.querySelector('input#user_password')
+  const passwordLength = passwordInput instanceof HTMLInputElement ? passwordInput.value.length : 0
+  return value.length >= passwordLength
+}
+
+const usePasswordConfirmationState = (formNode) => {
+  const [value, setValue] = useState('')
+  const [isValid, setIsValid] = useState(undefined)
+  const [errorMessage, setErrorMessage] = useState('isMandatory')
+  const [startValidating, setStartValidating] = useState(false)
+
+  const onChange = (value, event) => {
+    setValue(value)
+    const { hasError, errorMessage } = getPasswordConfirmationErrors(formNode)
+    setErrorMessage(errorMessage)
+    const initValidation = compareInputsLength(value)
+    const validationStarted = startValidating || isValid !== undefined
+    setStartValidating(!validationStarted ? initValidation : undefined)
+    setIsValid((initValidation || validationStarted) ? !hasError : undefined)
+  }
+
+  const onBlur = () => {
+    const errors = !!isFormValid(formNode)
+    setIsValid(!errors)
+  }
+
+  return {
+    value,
+    isValid,
+    errorMessage,
+    onChange,
+    onBlur
+  }
+}
+
+const ChangePassword = ({ lostPasswordToken, url, errors }: Props) => {
+  const formNode = document.querySelector('form')
+  const form = useFormState(formNode)
+  const password = usePasswordState()
```

Comment:
Intuitively, this should be something like: `usePasswordState({ value: '', isValid: false })`.

Or perhaps `const { value, isValid } = usePasswordState()`.